### PR TITLE
core: store blocks as doubly-linked list on Region [2/2]

### PR DESCRIPTION
**Pull request by superlopuh**
This migrates the blocks to be a doubly linked list.

### CODE REVIEW

File: xdsl/rewriter.py [hunk 1]
```diff
@@ -283,46 +283,34 @@ def insert_op_before(op: Operation, new_op: Operation):
     def move_region_contents_to_new_regions(region: Region) -> Region:
         """Move the region blocks to a new region."""
         new_region = Region()
-        for block in region.blocks:
```
Comment (math-fehr) [inline xdsl/rewriter.py:286]: Nice

**Review by math-fehr — APPROVED** [review]
Makes sense to me, though I think we *really* should abstract away the double linked list.
I'm pretty sure we can abstract a `DoubleLinkedList[T]` class that will contain almost all of the logic, and then with a bit of inheritance we can get back what you have (in both blocks and operations).
Otherwise, I'm fine merging it like it is

**Comment by superlopuh** [comment]
It would be nice to have such an abstraction, but I would say that this is something we want to hide from users, at least the mutability aspect. Let's chat about this at some point.

File: xdsl/ir/core.py [hunk 2]
```diff
@@ -1560,37 +1615,48 @@ def __getitem__(self, idx: int) -> Block: ...
...
+                        return b
+                raise IndexError
+        else:
+            # This is possible but would require a bit of work to handle complex slices
+            raise NotImplementedError("Indexing of RegionBlocks not yet implemented")
 
     def __len__(self):
-        return len(self._blocks)
+        i = 0
+        for _ in self:
```
Comment (georgebisbas) [inline xdsl/ir/core.py:1635]: why this way?

Comment (superlopuh) [reply xdsl/ir/core.py:1635]: as opposed to enumerating or `while (next := next.next) is not None`?

**Pull request merged**